Feature/io

lqcharska opened the pull request.

lqcharska committed:
DataAccessor - initial commit
 - load .dat and .hea file
 - data file is a vector of bytes
 - header file is a vector of strings

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>

lqcharska committed:
DataAccessor - remove dummy file

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>

bartek-krk commented on DataAccessor/DataAccessor.cpp:
```diff
@@ -0,0 +1,53 @@
+    loaded = false;
+    namespace fs = std::filesystem;
+
+    int splitPoint = filepath.find_last_of('/');
```
Będziemy odpalali ten kod zarówno na Windowsie jak i na MacOS/Linux. Zmień proszę "hardcodowany" separator na zależny od systemu:
```
#ifdef WIN32
#define OS_SEP '\\'
#else
#define OS_SEP '/'
#endif
```

lqcharska replied:
Zmieniłam do '/' i dorzuciłam żeby zamieniał z '\\' na '/'. Windows poprawnie interpretuje oba.

bartek-krk commented on DataAccessor/DataAccessor.cpp:
```diff
@@ -0,0 +1,53 @@
+    std::string fileNameWithoutExtension = fileNameWithExtension.substr(0, fileNameWithExtension.find_first_of('.'));
+
+    // Reading .dat file
+    std::string datFilePath = directoryPath + "/" + fileNameWithoutExtension + ".dat";
```
Zmień proszę separator również tutaj.

bartek-krk commented on DataAccessor/DataAccessor.cpp:
```diff
@@ -0,0 +1,53 @@
+    datBuffer = std::move(buffer1);
+
+    // Reading .hea file
+    std::string heaFilePath = directoryPath + "/" + fileNameWithoutExtension + ".hea";
```
Separator.

bartek-krk commented on DataAccessor/DataAccessor.h:
```diff
@@ -0,0 +1,56 @@
+     * 
+     * @return std::vector<unsigned char>& 
+     */
+    std::vector<unsigned char>& datBufferGet();
```
Pik `.dat` ma postać sekwencji `double`. Powinniśmy użyć `std::vector<double>` dokonując konwersji/parsowania.

bartek-krk commented on DataAccessor/DataAccessor.h:
```diff
@@ -0,0 +1,56 @@
+     * 
+     * @return std::vector<std::string>& 
+     */
+    std::vector<std::string>& heaBufferGet();
```
Możemy spróbować podzielić dane w plikach `.hea`. Znalazłem przykładową zawartość:
```
 >> cat nsr001.hea
nsr001 0 128 0
# Age: 64 Sex: F
```
Do ustalenia pozostaje, które dane z pliku będą przydatne.

bartek-krk reviewed (changes requested):
Ten moduł powinien zwracać gotowe dane dla innych modułów. Jeżeli będziemy chcieli wyświetlić w UI np. wiek z pliku `.hea` ta klasa powinna posiadać metodę `int getAge()` - nie powinniśmy zostawiać dowolności interpretacji tych nagłówków, ani szczegóły formatowania plików nie powinny być dostępne dla dalszych implementacji.

lqcharska committed:
DataAccessor - add functions from wfdb library.
DataAccessor can parse header and data file.

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>

lqcharska commented:
Na windowsie testowałam ładowanie rekordów za pomocą DataAccessora. 
Niech osoby z linuxem/mac'iem przetestują u siebie.

użycie: 
```
DataAccessor data;
data.load("C:/reco/100.hea");
auto signalInfo = data.signalInfoGet();
auto signalData = data.signalDataGet();
```

Funkcje z biblioteki wfdb przekopiowałam i przerobiłam, tak, żeby działały na windowsie, bo sama biblioteka (jak wcześniej zgłaszałam) ma jakiś problem na tym systemie.

lqcharska commented:
> Ten moduł powinien zwracać gotowe dane dla innych modułów. Jeżeli będziemy chcieli wyświetlić w UI np. wiek z pliku `.hea` ta klasa powinna posiadać metodę `int getAge()` - nie powinniśmy zostawiać dowolności interpretacji tych nagłówków, ani szczegóły formatowania plików nie powinny być dostępne dla dalszych implementacji.

Informacje o wieku, płci i NYHA class są dodane w plikach headerowych danych jako komentarz i biblioteka ich nie parsuje. Jeśli mają być parsowane, to nie możemy używać tej biblioteki, tylko należy użyć moich implementacji i zmodyfikować je o czytanie komentarzy

lqcharska committed:
DataAccessor - add comment parser

There are two expected comment styles:
 - "Age: yy  Sex: x  NYHA class: zzz"
 - "yy x"

Where:
 - yy is an age
 - x is a sex: F or M

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>

bartek-krk commented on DataAccessor/wfdb.h:
```diff
@@ -0,0 +1,76 @@
+#ifndef ECG_ANALYZERRRR_WFDB_H
+#define ECG_ANALYZERRRR_WFDB_H
+
+#include <wfdb_macros.h>
```
U mnie generuje to błąd - powinno być `#include "wfdb_macros.h"`.

MarcelPikula replied:
Do zmiany, plików szukamy wewnętrznie - stąd `""`.

MarcelPikula commented on DataAccessor/DataAccessor.h:
```diff
@@ -0,0 +1,110 @@
+     * 
+     * @return std::vector<std::vector<float>> 
+     */
+    std::vector<std::vector<float>> signalDataGet() const;
```
Typ zwracany przez tą metodę nie może mieć postać wektora wektorów. Znacznie skomplikuje to dalsze pracę modułów. Możemy to opakować w strukturę (lub inaczej ten problem rozwiązać).

lullusiek replied:
> W pliku `CMakeLists.txt` powinny zostać podane explicite wymagane kompilatory dla OSX:
> 
> ```
> set(CMAKE_CXX_COMPILER /usr/bin/g++)
> set(CMAKE_C_COMPILER  /usr/bin/gcc)
> ```
> 
> Jeżeli nie jest to koniecznie w Windowsie powyższa zmiana może zostać zacommentowana z adnotacją sugerującą odcommentowanie na OSX. Dodaj proszę stosowny punkt w `README.md`.

Ja nie zmieniam żadnego pliku CMake, więc dlaczego u mnie jest wymagane dodanie tego?
Jeśli dotyczy to głównego CMakeLists.txt to proponuję dodać w nowym PR, a ja się zrebasuje na niego jak wejdzie do maina.

MarcelPikula commented on DataAccessor/DataAccessor.cpp:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    // Checking .hea file
+    m_heaFilePath = directoryPath + m_separator + m_fileNameWithoutExtension + ".hea";
```
To zachowanie jest niebezpieczne. Nie możemy oczekiwać, że zawsze te pliki będą znajdowały się w jednym katalogu. Ustaliliśmy, że na wejście `DataAcessor` przyjdą dwa `std::string`: jeden do pliku `.dat` drugi do pliku `.hea`. Zmień proszę implementację tak aby uwzględnić nowy argument.

lqcharska replied:
Niestety biblioteka biblioteka wfdb (to co z niej skopiowałam) opiera się na tym, że plik `.hea` i `.dat` znajduje się w tym samym katalogu. Głównie dlatego, że w headerze jest informacja jak nazywa się plik z danymi (lub nazywają się, bo może być ich więcej). Więc tak na prawdę to powinniśmy wskazywać tylko na header.

bartek-krk reviewed (changes requested):
W pliku `CMakeLists.txt` powinny zostać podane explicite wymagane kompilatory dla OSX:
```
set(CMAKE_CXX_COMPILER /usr/bin/g++)
set(CMAKE_C_COMPILER  /usr/bin/gcc)
```
Jeżeli nie jest to koniecznie w Windowsie powyższa zmiana może zostać zacommentowana z adnotacją sugerującą odcommentowanie na OSX. Dodaj proszę stosowny punkt w `README.md`.

lqcharska commented:
> W pliku `CMakeLists.txt` powinny zostać podane explicite wymagane kompilatory dla OSX:
> 
> ```
> set(CMAKE_CXX_COMPILER /usr/bin/g++)
> set(CMAKE_C_COMPILER  /usr/bin/gcc)
> ```
> 
> Jeżeli nie jest to koniecznie w Windowsie powyższa zmiana może zostać zacommentowana z adnotacją sugerującą odcommentowanie na OSX. Dodaj proszę stosowny punkt w `README.md`.

Ja nie zmieniam żadnego pliku CMake, więc dlaczego u mnie jest wymagane dodanie tego?
Jeśli dotyczy to głównego CMakeLists.txt to proponuję dodać w nowym PR, a ja się zrebasuje na niego jak wejdzie do maina.

dadm2022 commented:
> > W pliku `CMakeLists.txt` powinny zostać podane explicite wymagane kompilatory dla OSX:
> > ```
> > set(CMAKE_CXX_COMPILER /usr/bin/g++)
> > set(CMAKE_C_COMPILER  /usr/bin/gcc)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Jeżeli nie jest to koniecznie w Windowsie powyższa zmiana może zostać zacommentowana z adnotacją sugerującą odcommentowanie na OSX. Dodaj proszę stosowny punkt w `README.md`.
> 
> Ja nie zmieniam żadnego pliku CMake, więc dlaczego u mnie jest wymagane dodanie tego? Jeśli dotyczy to głównego CMakeLists.txt to proponuję dodać w nowym PR, a ja się zrebasuje na niego jak wejdzie do maina.

To jest wymagane w systemach korzystających z kompilatora CLang. Uwzględnij uwagę Bartka - bez tego połowa zespołu nie będzie mogła korzystać z modułu I/O.

MarcelPikula commented:
> > W pliku `CMakeLists.txt` powinny zostać podane explicite wymagane kompilatory dla OSX:
> > ```
> > set(CMAKE_CXX_COMPILER /usr/bin/g++)
> > set(CMAKE_C_COMPILER  /usr/bin/gcc)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Jeżeli nie jest to koniecznie w Windowsie powyższa zmiana może zostać zacommentowana z adnotacją sugerującą odcommentowanie na OSX. Dodaj proszę stosowny punkt w `README.md`.
> 
> Ja nie zmieniam żadnego pliku CMake, więc dlaczego u mnie jest wymagane dodanie tego? Jeśli dotyczy to głównego CMakeLists.txt to proponuję dodać w nowym PR, a ja się zrebasuje na niego jak wejdzie do maina.

To jest wymagane w systemach korzystających z kompilatora CLang. Uwzględnij uwagę Bartka - bez tego połowa zespołu nie będzie mogła korzystać z modułu I/O.

MarcelPikula commented on DataAccessor/DataAccessor.cpp:
```diff
@@ -0,0 +1,182 @@
+                }
+
+                for (int j = 0; j < m_signalCount; j++) {
+                    dataBuffer[j].push_back(((float) sample[j] - baseline[j]) / gain[j]);
```
Jeżeli znamy ilość próbek dla poszczególnych sygnałów, proszę uwzględnij wywołanie `std::vector::reserve()` przed wypełnianiem wektora- spowoduje to zoptymalizowanie procesu wczytywania danych.

bartek-krk commented:
> > > W pliku `CMakeLists.txt` powinny zostać podane explicite wymagane kompilatory dla OSX:
> > > ```
> > > set(CMAKE_CXX_COMPILER /usr/bin/g++)
> > > set(CMAKE_C_COMPILER  /usr/bin/gcc)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Jeżeli nie jest to koniecznie w Windowsie powyższa zmiana może zostać zacommentowana z adnotacją sugerującą odcommentowanie na OSX. Dodaj proszę stosowny punkt w `README.md`.
> > 
> > 
> > Ja nie zmieniam żadnego pliku CMake, więc dlaczego u mnie jest wymagane dodanie tego? Jeśli dotyczy to głównego CMakeLists.txt to proponuję dodać w nowym PR, a ja się zrebasuje na niego jak wejdzie do maina.
> 
> To jest wymagane w systemach korzystających z kompilatora CLang. Uwzględnij uwagę Bartka - bez tego połowa zespołu nie będzie mogła korzystać z modułu I/O.

Modułu w tej postaci nie da się uruchomić - linker zgłasza błąd, który został wprowadzony w tym PR. Rozwiązanie problemu powyżej. Kod nie zostanie zmergowany bez tej zmiany.

lqcharska commented:
> W pliku CMakeLists.txt powinny zostać podane explicite wymagane kompilatory dla OSX:

W którym pliku `CMakeLists.txt`?

lqcharska committed:
DataAccessor - fix include.

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>

lqcharska committed:
DataAccessor - Create new Signal type with data and info.

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>

lqcharska committed:
DataAccessor - Add DataAccessor as library.

Signed-off-by: Laura Kucharska <lullusiek@gmail.com>
